iot/mqttsnippets/gateway_send_data_test.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	keyBytes, err := ioutil.ReadFile(publicKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	gateway := &cloudiot.Device{
+		Id: gatewayID,
+		Credentials: []*cloudiot.DeviceCredential{
+			{
+				PublicKey: &cloudiot.PublicKeyCredential{
+					Format: "RSA_X509_PEM",
+					Key:    string(keyBytes),
+				},
+			},
+		},
+		GatewayConfig: &cloudiot.GatewayConfig{
+			GatewayType:       "GATEWAY",
+			GatewayAuthMethod: gatewayAuthMethod,
+		},
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s/registries/%s", projectID, region, registryID)
+	response, err := client.Projects.Locations.Registries.Devices.Create(parent, gateway).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Fprintln(w, "Successfully created gateway:", gatewayID)
+
+	return response, nil
+}
+
+// bindDeviceToGateway creates an association between an existing device and gateway.
+func bindDeviceToGateway(w io.Writer, projectID string, region string, registryID string, gatewayID string, deviceID string) (*cloudiot.BindDeviceToGatewayResponse, error) {
+	client, err := getClient()
+	if err != nil {
+		return nil, err
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s/registries/%s", projectID, region, registryID)
+	bindRequest := &cloudiot.BindDeviceToGatewayRequest{
+		DeviceId:  deviceID,
+		GatewayId: gatewayID,
+	}
+
+	response, err := client.Projects.Locations.Registries.BindDeviceToGateway(parent, bindRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("BindDeviceToGateway: %v", err)
+	}
+
+	if response.HTTPStatusCode/100 != 2 {
+		return nil, fmt.Errorf("BindDeviceToGateway: HTTP status code not 2xx\n %v", response)
+	}
+
+	fmt.Fprintf(w, "Bound %s to %s", deviceID, gatewayID)
+
+	return response, nil
+}
+
+// unbindDeviceFromGateway unbinds a bound device from a gateway.
+func unbindDeviceFromGateway(w io.Writer, projectID string, region string, registryID string, gatewayID string, deviceID string) (*cloudiot.UnbindDeviceFromGatewayResponse, error) {
+	client, err := getClient()
+	if err != nil {
+		return nil, err
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s/registries/%s", projectID, region, registryID)
+	unbindRequest := &cloudiot.UnbindDeviceFromGatewayRequest{
+		DeviceId:  deviceID,
+		GatewayId: gatewayID,
+	}
+
+	response, err := client.Projects.Locations.Registries.UnbindDeviceFromGateway(parent, unbindRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("UnbindDeviceFromGateway error: %v", err)
+	}
+
+	if response.HTTPStatusCode/100 != 2 {
+		return nil, fmt.Errorf("UnbindDeviceFromGateway: HTTP status code not 2xx\n %v", response)
+	}
+
+	fmt.Fprintf(w, "Unbound %s from %s", deviceID, gatewayID)
+
+	return response, nil
+}
+
+func TestSendDataFromBoundDevice(t *testing.T) {
+	projectID := testutil.SystemTest(t).ProjectID
+
+	registryID := "golang-iot-test-registry"
+	gatewayID := createIDForTest("gateway")
+	deviceID := createIDForTest("device")
+
+	testutil.Retry(t, 1, 10*time.Second, func(r *testutil.R) {
```

Comment:
Might be better to have a retry for each part that could fail, rather than the end-to-end test. As is, if the last step fails, you have to redo all of the others.